lib/ansible/modules/crypto/openssh_cert.py

```diff
@@ -0,0 +1,542 @@
+                rmtree(temp_directory, ignore_errors=True)
+                proc = module.run_command([self.ssh_keygen, '-L', '-f', self.path])
+                self.cert_info = proc[1].split()
+                self.changed = True
+            except Exception as e:
+                try:
+                    self.remove()
+                    rmtree(temp_directory, ignore_errors=True)
+                except OSError as exc:
+                    if exc.errno != errno.ENOENT:
+                        raise CertificateError(exc)
+                    else:
+                        pass
+                module.fail_json(msg="%s" % to_native(e))
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+
+    def convertToDatetime(self, module, timestring):
+
+        if self.isRelative(timestring):
+            dispatched_time = re.findall("^([+\\-])((\\d+)[w])?((\\d+)[d])?((\\d+)[h])?((\\d+)[m])?((\\d+)[s])?$", timestring, re.I)
+            if not dispatched_time:
+                module.fail_json(msg="'%s' is not a valid time format." % timestring)
+            dispatched_time = dispatched_time[0]
+            if dispatched_time[0] == "+":
+                return datetime.utcnow() + timedelta(
+                    weeks=int('0' + dispatched_time[2]),
+                    days=int('0' + dispatched_time[4]),
+                    hours=int('0' + dispatched_time[6]),
+                    minutes=int('0' + dispatched_time[8]),
+                    seconds=int('0' + dispatched_time[10]))
+            else:
+                return datetime.utcnow() - timedelta(
+                    weeks=int('0' + dispatched_time[2]),
+                    days=int('0' + dispatched_time[4]),
+                    hours=int('0' + dispatched_time[6]),
+                    minutes=int('0' + dispatched_time[8]),
+                    seconds=int('0' + dispatched_time[10]))
+        else:
+            formats = ["%Y:%m:%d", "%Y:%m:%d %H:%M:%S", "%Y:%m:%d:%H:%M:%S", "%Y-%m-%dT%H:%M:%S"]
+            for fmt in formats:
+                try:
+                    return datetime.strptime(timestring, fmt)
+                except:
+                    pass
+            module.fail_json(msg="'%s' is not a valid time format" % timestring)
+
+    def isRelative(self, timestr):
+        if timestr.startswith("+") or timestr.startswith("-"):
+            return True
+        return False
+
+    def isValid(self, module, perms_required=True):
+
+        def _check_state():
+            return os.path.exists(self.path)
+
+        if _check_state():
+            proc = module.run_command([self.ssh_keygen, '-L', '-f', self.path], environ_update=dict(TZ="UTC"))
+            self.cert_info = proc[1].split()
+            principals = re.findall("(?<=Principals:)(.*)(?=Critical)", proc[1], re.S)[0].split()
```

Comment:
What will happen if `ssh-keygen` outputs something different (like not a valid file)? This will probably break.

Reply:
(In fact, `run_command()` might already die when `ssh-keygen` returns a non-zero return code. I think there's a flag for `run_command()` to force it to not to do that.)

Reply:
I would add `check_rc=False` anyway, to make clearer your intention that you accept non-zero return codes.